### ISSUE TITLE
Breaking changes to fullscreen operation

### ISSUE THREAD

**Issue by peppy** [libsdl-org/SDL#6033]
We recently updated our .NET binaries from https://github.com/libsdl-org/SDL/commit/4f19421ed8e24aef9075792a1c33f38b42f949f1 (previously aeeccb5d16207b2b18afe6bcde12efe52b0e09b4) and noticed two changes/regressions in behaviour.

- `SDL_WINDOWEVENT_ENTER` is no longer fired after returning from alt-tabbing a fullscreen window Previously this would always fire in this scenario.
- `SDL_GetWindowSize` returns zero when running fullscreen and minimised, where it didn't before. This caused hard crashes for us as we were using it to calculate [screen scale](https://github.com/ppy/osu-framework/pull/5344/files).

Apologies if either of these issues has since been resolved. Figure it's easier to report this initially and provide more detail as/if required. Let me know if you need isolated reproduction and I can look into that.

**Comment by icculus** [comment]
Which platform is this happening on? Windows? Mac? X11? All of them?!

**Comment by icculus** [comment]
Putting this in 2.24, but we're down to the wire here.

**Comment by peppy** [comment]
This is specifically on Windows, and we have not seen any reproduction on other platforms.

**Comment by slouken** [comment]
I modified testsprite2.c as follows, and then ran with `--fullscreen` and pressed Ctrl-Z to minimize the window and then clicked in the taskbar to restore the window.
```diff
diff --git a/test/testsprite2.c b/test/testsprite2.c
index a91ddd86e..aab278748 100644
--- a/test/testsprite2.c
+++ b/test/testsprite2.c
@@ -404,6 +404,16 @@ loop()

     /* Check for events */
     while (SDL_PollEvent(&event)) {
+        if (event.type == SDL_WINDOWEVENT) {
+            if (event.window.event == SDL_WINDOWEVENT_ENTER) {
+                SDL_Log("WINDOW ENTER!\n");
+            }
+            if (event.window.event == SDL_WINDOWEVENT_MINIMIZED || event.window.event == SDL_WINDOWEVENT_RESTORED) {
+                int w, h;
+                SDL_GetWindowSize(state->windows[0], &w, &h);
+                SDL_Log("WINDOW %s: %dx%d\n", (event.window.event == SDL_WINDOWEVENT_MINIMIZED) ? "MINIMIZED" : "RESTORED", w, h);
+            }
+        }
         SDLTest_CommonEvent(state, &event, &done);
     }
     for (i = 0; i < state->num_windows; ++i) {

```
Here was my output:
```
INFO: WINDOW ENTER!
INFO: WINDOW MINIMIZED: 237x39
INFO: WINDOW RESTORED: 640x480
INFO: WINDOW ENTER!
```
- Are you getting the same results with testsprite2?
- Do you have anything special like a multi-monitor setup?
- Are you able to bisect and see where this broke for you?

**Comment by peppy** [comment]
I can reproduce the correct results using the SDL-side test, but it definitely doesn't provide the same behaviour in our game. I'm in the process of bisecting as I think that will be the quickest way to find the root cause (already tried disabling all hints / window flags we have set to no avail).

**Comment by peppy** [comment]
```
❯ git bisect bad
ab81a559f43abc0858c96788f8e00bbb352287e8 is the first bad commit
commit ab81a559f43abc0858c96788f8e00bbb352287e8
```

I'll do some further testing / investigation but dropping this here in case you figure something out before I can.

**Comment by slouken** [comment]
@ericwa, can you look at this? It's one of 2 bugs holding up 2.24.0 release.

**Comment by slouken** [comment]
@peppy, what happens if you hardcode dpi to 96 in `WIN_AdjustWindowRectWithStyle()`?

This comment looks relevant:
```
       - Bug workaround: when leaving exclusive fullscreen, the cached DPI and window DPI reported
         by GetDpiForWindow will be wrong, and would cause windows shrinking slightly when
         going from exclusive fullscreen to windowed on a HighDPI monitor with scaling if we used them.
```

**Comment by peppy** [comment]
To confirm, do you mean like this?

<img width="486" alt="Parallels Desktop 2022-08-12 at 15 04 47@2x" src="https://user-images.githubusercontent.com/191335/184384024-a0d0f007-1245-4e64-a296-6772971078bb.png">

**Comment by ericwa** [comment]
Cool, looks like you found a fix @slouken - I was just getting started reading the comments and trying to repro this.

We should watch out for issues stemming from `WIN_GetDrawableSize` as that is one thing that changed in https://github.com/libsdl-org/SDL/commit/ab81a559f43abc0858c96788f8e00bbb352287e8 - code that was previously calling `SDL_GetWindowSize` and assuming the SDL window `->w` and `->h`  were in pixels, is now calling `WIN_GetDrawableSize` (which is just a wrapper around `GetClientRect`.) So that's a change from reading cached sizes to a win32 API call.

**Comment by peppy** [comment]
I can confirm all reported regressions are fixed on be0cf25. Thanks!

**Comment by slouken** [comment]
Thanks! :)
